File: q23.c
#include <stdio.h>
#include <stdlib.h>

// Node structure for the binary search tree
struct Node {
    int data;
    struct Node *left, *right;
};

// Function to create a new node
struct Node* newNode(int data) {
    struct Node* node = (struct Node*)malloc(sizeof(struct Node));
    node->data = data;
    node->left = node->right = NULL;
    return node;
}

// Function to insert a new node with a given key into BST
struct Node* insert(struct Node* node, int key) {
    if (node == NULL) return newNode(key);
    if (key < node->data)
        node->left = insert(node->left, key);
    else if (key > node->data)
        node->right = insert(node->right, key);
    return node;
}

// Function to perform inorder traversal of BST
void inorder(struct Node* root) {
    if (root != NULL) {
        inorder(root->left);
        printf("%d ", root->data);
        inorder(root->right);
    }
}

// Function to search for a key in BST
struct Node* search(struct Node* root, int key) {
    if (root == NULL || root->data == key)
        return root;
    if (root->data < key)
        return search(root->right, key);
    return search(root->left, key);
}

int main() {
    struct Node* root = NULL;
    int keys[] = {5, 3, 8, 2, 4, 7, 10};
    int n = sizeof(keys) / sizeof(keys[0]);

    // Insert keys into BST
    for (int i = 0; i < n; i++)
        root = insert(root, keys[i]);

    // Search for key 7
    struct Node* node = search(root, 7);
    printf("Search path: ");
    while (node != NULL) {
        printf("%d ", node->data);
        if (node->data < 7)
            node = node->right;
        else if (node->data > 7)
            node = node->left;
        else
            break;
    }
    printf("\n");

    return 0;
}

File: q23.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def insert(root, key):
    if root is None:
        return Node(key)
    if key < root.data:
        root.left = insert(root.left, key)
    elif key > root.data:
        root.right = insert(root.right, key)
    return root

def inorder(root):
    if root:
        inorder(root.left)
        print(root.data, end=" ")
        inorder(root.right)

def search(root, key):
    if root is None or root.data == key:
        return root
    if root.data < key:
        return search(root.right, key)
    return search(root.left, key)

# Construct the binary search tree
keys = [5, 3, 8, 2, 4, 7, 10]
root = None
for key in keys:
    root = insert(root, key)

# Search for key 7 and print the search path
node = search(root, 7)
print("Search path:", end=" ")
while node:
    print(node.data, end=" ")
    if node.data < 7:
        node = node.right
    elif node.data > 7:
        node = node.left
    else:
        break
